src/internal/observable/dom/animationFrames.ts

```diff
@@ -0,0 +1,214 @@
+ * source$.subscribe(x => console.log(x));
+ * ```
+ *
+ * @param timestampProvider An object with a `now` method that provides a numeric timestamp
+ */
+export function animationFrames(timestampProvider: TimestampProvider = Date) {
+  return timestampProvider === Date ? DEFAULT_ANIMATION_FRAMES : animationFramesFactory(timestampProvider);
+}
+
+/**
+ * Does the work of creating the observable for `animationFrames`.
+ * @param timestampProvider The timestamp provider to use to create the observable
+ */
+function animationFramesFactory(timestampProvider: TimestampProvider) {
+  return new Observable<number>(subscriber => {
+    subscriberState.push(subscriber, timestampProvider, timestampProvider.now());
+    startAnimationLoop();
+    return () => {
+      if (isCurrentlyNotifying) {
+        // The `animate` loop is currently firing. We need to wait before we
+        // remove the subscriber from the list.
+        subscribersToRemoveAfterRun.push(subscriber);
+      } else {
+        removeSubscriber(subscriber);
+      }
+    };
+  });
+}
+
+/**
+ * In the common case, where `Date` is passed to `animationFrames`, we use
+ * this shared observable to reduce memory pressure.
+ */
+const DEFAULT_ANIMATION_FRAMES = animationFramesFactory(Date);
+
+/**
+ * Removes a subscriber, and its accompanying data, from the list that gets notified when an animation frame has fired.
+ *
+ * NOTE: This implementation is relying on the fact that `subscriber` will always be a different instance as it is passed
+ * into the observable's initialization function (passed to the Observable ctor).
+ *
+ * @param subscriber the subscriber to remove from the list of subscribers to notify
+ */
+function removeSubscriber(subscriber: Subscriber<any>) {
+  const index = subscriberState.indexOf(subscriber);
+  if (index >= 0) {
+    // Remove the subscriber and its timestampProvider
+    subscriberState.splice(index, 3);
+    if (subscriberState.length === 0) {
+      stopAnimationLoop();
+    }
+  }
+}
+
+/**
+ * The currently scheduled animation frame id.
+ */
+let scheduledAnimationId = 0;
+
+/**
+ * If `true`, the `animate` loop is currently notifying the subscribers.
+ *
+ * We have this so we can see if unsubscription should defer the removal of subscribers from
+ * the inner list. This is okay, because subscribers that are already unsubscribed will not notify,
+ * and it saves us from needing to copy the array of subscribers prior to the run loop.
+ */
+let isCurrentlyNotifying = false;
+
+/**
+ * Starts the animation frame `animate` loop, if necessary.
+ * Idempotent. If called and it's already scheduled to start, it will not reschedule or cancel.
+ */
+function startAnimationLoop() {
+  if (scheduledAnimationId === 0) {
+    scheduledAnimationId = requestAnimationFrame(animate);
+  }
+}
+
+/**
+ * Executes notification of all subscribers, then reschedules itself.
+ * Do not call directly. This is the "run loop".
+ */
+function animate() {
```

Comment:
This is variously referred to as the `run` loop and the `animate` loop. Maybe pick one term for it and use that term only?